sdks/python/apache_beam/transforms/core.py

```diff
@@ -1387,6 +1388,47 @@ def partition_for(self, element, num_partitions, *args, **kwargs):
     return self._fn(element, num_partitions, *args, **kwargs)
 
 
+def _get_function_body_without_inners(func):
+  source_lines = inspect.getsourcelines(func)[0]
```

Comment:
I think analyzing bytecode (see `dis.get_instructions()`) would be more reliable that src analysis. Have you considered that?

Reply:
I tried this as my first idea. It does not work:
```
def bar():
    yield "bar"
list(dis.get_instructions(bar))
```
This generates:
```
[Instruction(opname='LOAD_CONST', opcode=100, arg=1, argval='bar', argrepr="'bar'", offset=0, starts_line=2, is_jump_target=False),
 Instruction(opname='YIELD_VALUE', opcode=86, arg=None, argval=None, argrepr='', offset=2, starts_line=None, is_jump_target=False),
 Instruction(opname='POP_TOP', opcode=1, arg=None, argval=None, argrepr='', offset=4, starts_line=None, is_jump_target=False),
 Instruction(opname='LOAD_CONST', opcode=100, arg=0, argval=None, argrepr='None', offset=6, starts_line=None, is_jump_target=False),
 Instruction(opname='RETURN_VALUE', opcode=83, arg=None, argval=None, argrepr='', offset=8, starts_line=None, is_jump_target=False)]
```
`RETURN_VALUE` is always there no matter the function contains yield or return. So I decided to use this simple way by just checking the source code.

Reply:
`total_yield` and `tax_return` could  trigger false positives here. I took a closer look, I am no longer sure unconditionally failing  the pipeline is the right thing to do here, as we are fighting against Python language rules. I think we can print a warning potentially, and update the documentation to recommend `yield from` instead of return for iterables. We should do the due diligence to make sure `yield from` achieves the desired result though to make sure there are no surprises.

Reply:
By checking the lines, we can capture these cases now. I also updated the tests. To be safe, I changed to use the logging warning instead of raising RuntimeError. The doc was updated by https://github.com/apache/beam/pull/25060. I also updated the warning messages with the actionable suggestions.

Reply:
> The doc was updated by #25060

we could recommend yield from instead of return for iterables, after checking that yield from works well.

Reply:
Done. and also log it with info.